File: docs/language-lab/python/test_lab.py
import unittest

from lab_kernel import (
    Atom,
    Bucket,
    Effect,
    Kind,
    Node,
    Out,
    Row,
    SELF,
    Verdict,
    by,
    decode,
    decode_atom,
    encode,
    encode_atom,
    exact,
    fact_id,
    frame,
    make_fact,
    span,
)
from lab_runtime import OutLink, WireDecoder, cycle, outbox, pump, wire_message


def offer(role, scope, target, value=None):
    return Atom(Kind.OFFER, role, scope, target, value)


def need(role, scope, target, effect):
    return Atom(Kind.NEED, role, scope, target, effect=effect)


class DemoRoot:
    def extract(self, fact):
        return fact.tag != b"courier"

    def project(self, fact, context):
        if fact.tag == b"invalid":
            return Out(Verdict.INVALID)
        if fact.tag == b"courier" and by(context, b"shipped"):
            return Out(Verdict.REAP)
        if fact.tag == b"clock" and not by(context, b"now"):
            return Out()
        if fact.tag in (b"pass", b"courier", b"clock"):
            return Out(offers=tuple(atom for atom in fact.atoms if atom.kind is Kind.OFFER))
        return None


class LanguageLabTest(unittest.TestCase):
    def test_canonical_round_trip_and_golden_id(self):
        atoms = (
            offer(b"result", b"s", SELF, b"ok"),
            need(b"dep", b"s", exact(b"key"), Effect.REQUIRE),
        )
        fact = make_fact(
            b"pass",
            atoms[1],
            atoms[0],
            atoms[1],
        )
        self.assertEqual(len(fact.atoms), 2)
        blob = encode(fact)
        self.assertEqual(decode(blob), fact)
        self.assertEqual(
            fact_id(fact).hex(),
            "33a234f18d975af511b7648e6199ac1db55521a60b811e1478e57fe16943b8c7",
        )
        self.assertIsNone(Node(DemoRoot()).admit(blob[:-1]))
        reversed_fact = frame(fact.tag) + b"".join(
            frame(encode_atom(atom)) for atom in reversed(fact.atoms)
        )
        self.assertIsNone(Node(DemoRoot()).admit(reversed_fact))
        noncanonical_range = frame(
            bytes((Kind.OFFER, Effect.NONE, 2)), b"r", b"s", b"x", b"x"
        )
        with self.assertRaises(ValueError):
            decode_atom(noncanonical_range)

    def test_bucket_matches_point_and_range_both_directions(self):
        point = Row(b"p", 0, offer(b"r", b"s", exact(b"m")))
        ranged = Row(b"r", 0, offer(b"r", b"s", span(b"a", b"z")))
        bucket = Bucket()
        bucket.add(point)
        bucket.add(ranged)
        self.assertEqual(set(bucket.matching(exact(b"m"))), {point, ranged})
        self.assertEqual(bucket.matching(span(b"l", b"n")), [point])
        self.assertEqual(bucket.matching(span(b"a", b"z")), [point])

    def test_require_parks_then_offer_wakes_and_promotes(self):
        node = Node(DemoRoot())
        dependent = make_fact(
            b"pass",
            need(b"dep", b"s", exact(b"key"), Effect.REQUIRE),
            offer(b"result", b"s", SELF, b"yes"),
        )
        dep_id = node.admit(encode(dependent))
        node.run()
        self.assertEqual(node.memo[dep_id], Verdict.PARKED)
        self.assertEqual(node.watched(b"result", b"s"), [])
        provider = make_fact(b"pass", offer(b"dep", b"s", exact(b"key"), b"ready"))
        node.admit(encode(provider))
        node.run()
        self.assertEqual(node.memo[dep_id], Verdict.VALID)
        self.assertEqual([row.owner for row in node.watched(b"result", b"s")], [dep_id])

    def test_suppress_withdraws_and_evicts_whole_owner(self):
        node = Node(DemoRoot())
        victim = make_fact(
            b"pass",
            need(b"dead", b"s", SELF, Effect.SUPPRESS),
            offer(b"live", b"s", SELF),
        )
        victim_id = node.admit(encode(victim))
        node.run()
        killer = make_fact(b"pass", offer(b"dead", b"s", exact(victim_id)))
        node.admit(encode(killer))
        node.run()
        self.assertNotIn(victim_id, node.facts)
        self.assertNotIn(victim_id, node.durable)
        self.assertEqual(node.watched(b"live", b"s"), [])

        precedence_victim = make_fact(
            b"pass",
            need(b"dead-priority", b"s", SELF, Effect.SUPPRESS),
            need(b"missing", b"s", exact(b"nothing"), Effect.REQUIRE),
            offer(b"never", b"s", SELF),
        )
        precedence_id = fact_id(precedence_victim)
        precedence_killer = make_fact(
            b"pass", offer(b"dead-priority", b"s", exact(precedence_id))
        )
        node.admit(encode(precedence_killer))
        node.run()
        node.admit(encode(precedence_victim))
        node.run()
        self.assertNotIn(precedence_id, node.facts)
        self.assertEqual(node.watched(b"never", b"s"), [])

    def test_clock_watch_and_bound(self):
        node = Node(DemoRoot())
        clocked = make_fact(
            b"clock",
            need(b"now", b"clock", span((100).to_bytes(8, "big"), b"\xff" * 8), Effect.WATCH),
            offer(b"ready", b"s", SELF),
        )
        node.admit(encode(clocked))
        node.turn(99, bound=1)
        self.assertEqual(node.watched(b"ready", b"s"), [])
        node.turn(100, bound=1)
        self.assertEqual(len(node.watched(b"ready", b"s")), 1)
        node.turn(101, bound=1)
        now_rows = node.watched(b"now", b"clock")
        self.assertEqual(len(now_rows), 1)
        self.assertEqual(now_rows[0].atom.target.lo, (101).to_bytes(8, "big"))
        node.admit(encode(make_fact(b"pass", offer(b"a", b"s", SELF))))
        node.admit(encode(make_fact(b"pass", offer(b"b", b"s", SELF))))
        node.turn(bound=1)
        self.assertEqual(len(node.frontier), 1)

    def test_inline_pump_then_shipped_reaps(self):
        node = Node(DemoRoot())
        courier = make_fact(
            b"courier",
            offer(b"send", b"outbox", exact(b"peer"), b"hello"),
            need(b"shipped", b"wire", SELF, Effect.WATCH),
        )
        courier_id = fact_id(courier)
        cycle(node, [encode(courier)], 1)
        got = []
        fired = pump(
            node,
            lambda cid: (b"127.0.0.1:9", b"secret") if cid == b"peer" else None,
            lambda cid, address, secret, inners: got.extend(inners) or len(inners),
            set(),
        )
        self.assertEqual(got, [b"hello"])
        self.assertEqual(fired, {courier_id})
        node.admit(encode(make_fact(b"pass", offer(b"backlog-a", b"s", SELF))))
        node.admit(encode(make_fact(b"pass", offer(b"backlog-b", b"s", SELF))))
        cycle(node, [], 2, tuple(fired), bound=1)
        self.assertIn(courier_id, node.facts)
        redelivered = []
        self.assertEqual(
            pump(
                node,
                lambda cid: (b"127.0.0.1:9", b"secret") if cid == b"peer" else None,
                lambda _cid, _address, _secret, inners: redelivered.extend(inners) or len(inners),
                fired,
            ),
            set(),
        )
        self.assertEqual(redelivered, [])
        cycle(node, [], 3, tuple(fired), bound=1)
        cycle(node, [], 4, tuple(fired), bound=1)
        self.assertNotIn(courier_id, node.facts)
        self.assertEqual(outbox(node), [])

    def test_reference_dedup_and_short_delivery_retry(self):
        node = Node(DemoRoot())
        first = make_fact(b"pass", offer(b"data", b"s", SELF, b"one"))
        second = make_fact(b"pass", offer(b"data", b"s", SELF, b"two"))
        first_id, second_id = fact_id(first), fact_id(second)
        node.admit(encode(first))
        node.admit(encode(second))
        node.run()
        def shipper(round_id):
            return make_fact(
                b"courier",
                offer(b"ship", b"outbox", exact(b"peer"), frame(first_id, second_id)),
                offer(b"round", b"test", SELF, round_id),
                need(b"shipped", b"wire", SELF, Effect.WATCH),
            )

        first_shipper = shipper(b"one")
        first_shipper_id = fact_id(first_shipper)
        cycle(node, [encode(first_shipper)], 1)
        sent = {}
        got = []
        fired = pump(node, lambda _: (b"a", None), lambda *args: got.extend(args[-1][:1]) or 1, set(), sent)
        self.assertEqual(got, [encode(first)])
        self.assertEqual(sent[b"peer"], {first_id})
        self.assertEqual(fired, {first_shipper_id})
        cycle(node, [], 2, tuple(fired))
        self.assertNotIn(first_shipper_id, node.facts)

        second_shipper = shipper(b"two")
        cycle(node, [encode(second_shipper)], 3)
        got.clear()
        pump(node, lambda _: (b"a", None), lambda *args: got.extend(args[-1]) or len(args[-1]), set(), sent)
        self.assertEqual(got, [encode(second)])

    def test_fragmented_wire_and_bounded_output(self):
        decoder = WireDecoder()
        wire = wire_message(0, b"hello") + wire_message(1, b"world")
        self.assertEqual(decoder.feed(wire[:3]), [])
        self.assertEqual(decoder.feed(wire[3:8]), [])
        self.assertEqual(decoder.feed(wire[8:]), [(0, b"hello"), (1, b"world")])
        link = OutLink(len(wire))
        self.assertTrue(link.enqueue(0, b"hello"))
        self.assertTrue(link.enqueue(1, b"world"))
        self.assertFalse(link.enqueue(1, b"overflow"))
        self.assertEqual(link.take(-1), b"")
        self.assertEqual(link.pending, len(wire))
        self.assertEqual(link.take(3) + link.take(10_000), wire)
        self.assertEqual(link.pending, 0)


if __name__ == "__main__":
    unittest.main()

File: language-lab/elixir/test_lab.exs
Code.require_file("kernel.exs", __DIR__)
Code.require_file("runtime.exs", __DIR__)

ExUnit.start()

defmodule TinyP2P.LanguageLab.DemoRoot do
  @behaviour TinyP2P.LanguageLab.Root

  alias TinyP2P.LanguageLab.{Kernel, Out}

  @impl true
  def extract(fact), do: fact.tag != "courier"

  @impl true
  def project(fact, context) do
    cond do
      fact.tag == "invalid" ->
        %Out{verdict: :invalid}

      fact.tag == "courier" and Kernel.by(context, "shipped") != [] ->
        %Out{verdict: :reap}

      fact.tag == "clock" and Kernel.by(context, "now") == [] ->
        %Out{}

      fact.tag in ["pass", "courier", "clock"] ->
        %Out{offers: Enum.filter(fact.atoms, &(&1.kind == :offer))}

      true ->
        nil
    end
  end
end

defmodule TinyP2P.LanguageLabTest do
  use ExUnit.Case, async: true

  alias TinyP2P.LanguageLab.{Atom, Bucket, Kernel, OutLink, Row, Runtime, WireDecoder}
  alias TinyP2P.LanguageLab.DemoRoot

  test "canonical construction, strict round trip, golden id, and malformed rejection" do
    result = offer("result", "s", :self, "ok")
    dependency = need("dep", "s", Kernel.exact("key"), :require)
    fact = Kernel.make_fact("pass", [result, dependency, result])
    reordered = Kernel.make_fact("pass", [dependency, result])

    assert fact == reordered
    assert length(fact.atoms) == 2

    blob = Kernel.encode(fact)
    assert {:ok, ^fact} = Kernel.decode(blob)

    assert fact |> Kernel.fact_id() |> Base.encode16(case: :lower) ==
             "33a234f18d975af511b7648e6199ac1db55521a60b811e1478e57fe16943b8c7"

    node = Kernel.new(DemoRoot)
    assert {^node, nil} = Kernel.admit(node, binary_part(blob, 0, byte_size(blob) - 1))

    reversed =
      Kernel.frame([fact.tag]) <>
        (fact.atoms
         |> Enum.reverse()
         |> Enum.map(fn atom -> Kernel.frame([Kernel.encode_atom(atom)]) end)
         |> IO.iodata_to_binary())

    assert {^node, nil} = Kernel.admit(node, reversed)

    duplicate =
      Kernel.frame([fact.tag]) <>
        Kernel.frame([Kernel.encode_atom(hd(fact.atoms))]) <>
        Kernel.frame([Kernel.encode_atom(hd(fact.atoms))])

    assert {^node, nil} = Kernel.admit(node, duplicate)

    degenerate_range = Kernel.frame([<<0, 2, 2>>, "r", "s", "x", "x"])
    malformed = Kernel.frame(["pass"]) <> Kernel.frame([degenerate_range])
    assert {:error, :invalid_fact} = Kernel.decode(malformed)

    extra_part = Kernel.frame([<<0, 2, 0>>, "r", "s", "x", "value", "extra"])
    assert {:error, :invalid_atom} = Kernel.decode_atom(extra_part)

    reserved_offer = Kernel.frame([<<1, 0, 0>>, <<0, "private">>, "s", "x"])
    assert {:error, :invalid_atom} = Kernel.decode_atom(reserved_offer)
  end

  test "exact and range buckets match in both point/range directions" do
    point = %Row{
      owner: "p",
      timestamp: 0,
      atom: offer("r", "s", Kernel.exact("m"))
    }

    ranged = %Row{
      owner: "r",
      timestamp: 0,
      atom: offer("r", "s", Kernel.span("a", "z"))
    }

    bucket = %Bucket{} |> Bucket.add(point) |> Bucket.add(ranged)

    assert MapSet.new(Bucket.matching(bucket, Kernel.exact("m"))) ==
             MapSet.new([point, ranged])

    assert Bucket.matching(bucket, Kernel.span("l", "n")) == [point]
    assert Bucket.matching(bucket, Kernel.span("a", "z")) == [point]
    refute Kernel.covers(Kernel.span("a", "z"), Kernel.span("l", "n"))
  end

  test "Require parks, then a matching validated offer wakes and promotes" do
    dependent =
      Kernel.make_fact("pass", [
        need("dep", "s", Kernel.exact("key"), :require),
        offer("result", "s", :self, "yes")
      ])

    {node, dependent_id} = Kernel.admit(Kernel.new(DemoRoot), Kernel.encode(dependent))
    node = Kernel.run(node)

    assert node.memo[dependent_id] == :parked
    assert Kernel.watched(node, "result", "s") == []

    provider =
      Kernel.make_fact("pass", [offer("dep", "s", Kernel.exact("key"), "ready")])

    {node, _provider_id} = Kernel.admit(node, Kernel.encode(provider))
    node = Kernel.run(node)

    assert node.memo[dependent_id] == :valid
    assert Enum.map(Kernel.watched(node, "result", "s"), & &1.owner) == [dependent_id]
  end

  test "Suppress wins over a missing Require after withdrawal and evicts the whole owner" do
    provider =
      Kernel.make_fact("pass", [
        need("gone", "s", :self, :suppress),
        offer("dep", "s", Kernel.exact("key"), "ready")
      ])

    provider_id = Kernel.fact_id(provider)

    victim =
      Kernel.make_fact("pass", [
        need("dep", "s", Kernel.exact("key"), :require),
        need("dead", "s", :self, :suppress),
        offer("live", "s", :self)
      ])

    victim_id = Kernel.fact_id(victim)
    {node, ^provider_id} = Kernel.admit(Kernel.new(DemoRoot), Kernel.encode(provider))
    {node, ^victim_id} = Kernel.admit(node, Kernel.encode(victim))
    node = Kernel.run(node)

    assert Enum.map(Kernel.watched(node, "live", "s"), & &1.owner) == [victim_id]

    provider_killer =
      Kernel.make_fact("pass", [offer("gone", "s", Kernel.exact(provider_id))])

    {node, _killer_id} = Kernel.admit(node, Kernel.encode(provider_killer))
    node = Kernel.run(node)

    refute Map.has_key?(node.facts, provider_id)
    assert node.memo[victim_id] == :parked
    assert Kernel.watched(node, "live", "s") == []

    victim_killer = Kernel.make_fact("pass", [offer("dead", "s", Kernel.exact(victim_id))])
    {node, _killer_id} = Kernel.admit(node, Kernel.encode(victim_killer))
    node = Kernel.run(node)

    refute Map.has_key?(node.facts, victim_id)
    refute Map.has_key?(node.memo, victim_id)
    refute Map.has_key?(node.durable, victim_id)
    assert Kernel.watched(node, "live", "s") == []
  end

  test "Watch reprojects from clock signals and turns leave bounded work queued" do
    clocked =
      Kernel.make_fact("clock", [
        need(
          "now",
          "clock",
          Kernel.span(<<100::unsigned-big-64>>, :binary.copy(<<255>>, 8)),
          :watch
        ),
        offer("ready", "s", :self)
      ])

    {node, _clocked_id} = Kernel.admit(Kernel.new(DemoRoot), Kernel.encode(clocked))
    node = Kernel.turn(node, 99, [], 1)
    assert Kernel.watched(node, "ready", "s") == []

    node = Kernel.turn(node, 100, [], 1)
    assert length(Kernel.watched(node, "ready", "s")) == 1

    first = Kernel.make_fact("pass", [offer("a", "s", :self)])
    second = Kernel.make_fact("pass", [offer("b", "s", :self)])
    {node, _first_id} = Kernel.admit(node, Kernel.encode(first))
    {node, _second_id} = Kernel.admit(node, Kernel.encode(second))
    node = Kernel.turn(node, nil, [], 1)

    assert :queue.len(node.frontier) == 1
  end

  test "inline courier pump fires, then the shipped signal Reaps it" do
    courier =
      Kernel.make_fact("courier", [
        offer("send", "outbox", Kernel.exact("peer"), "hello"),
        need("shipped", "wire", :self, :watch)
      ])

    courier_id = Kernel.fact_id(courier)
    node = Runtime.cycle(Kernel.new(DemoRoot), [Kernel.encode(courier)], 1)
    parent = self()

    route = fn
      "peer" -> {"127.0.0.1:9", "secret"}
      _other -> nil
    end

    deliver = fn cid, address, secret, inners ->
      send(parent, {:delivered, cid, address, secret, inners})
      length(inners)
    end

    {fired, _sent} = Runtime.pump(node, route, deliver, MapSet.new())

    assert fired == MapSet.new([courier_id])
    assert_receive {:delivered, "peer", "127.0.0.1:9", "secret", ["hello"]}

    node = Runtime.cycle(node, [], 2, MapSet.to_list(fired))
    refute Map.has_key?(node.facts, courier_id)
    assert Runtime.outbox(node) == []
  end

  test "by-reference pump deduplicates the delivered prefix and retries its tail" do
    first = Kernel.make_fact("pass", [offer("data", "s", :self, "one")])
    second = Kernel.make_fact("pass", [offer("data", "s", :self, "two")])
    first_id = Kernel.fact_id(first)
    second_id = Kernel.fact_id(second)

    {node, ^first_id} = Kernel.admit(Kernel.new(DemoRoot), Kernel.encode(first))
    {node, ^second_id} = Kernel.admit(node, Kernel.encode(second))
    node = Kernel.run(node)

    shipper =
      Kernel.make_fact("courier", [
        offer("ship", "outbox", Kernel.exact("peer"), Kernel.frame([first_id, second_id])),
        need("shipped", "wire", :self, :watch)
      ])

    shipper_id = Kernel.fact_id(shipper)
    node = Runtime.cycle(node, [Kernel.encode(shipper)], 1)
    parent = self()

    deliver_one = fn _cid, _address, _secret, inners ->
      send(parent, {:first_batch, Enum.take(inners, 1)})
      1
    end

    {fired, sent} =
      Runtime.pump(node, fn _cid -> {"a", nil} end, deliver_one, MapSet.new(), %{})

    assert fired == MapSet.new([shipper_id])
    assert_receive {:first_batch, [first_blob]}
    assert first_blob == Kernel.encode(first)
    assert sent["peer"] == MapSet.new([first_id])

    deliver_rest = fn _cid, _address, _secret, inners ->
      send(parent, {:second_batch, inners})
      length(inners)
    end

    {_fired, sent} =
      Runtime.pump(node, fn _cid -> {"a", nil} end, deliver_rest, MapSet.new(), sent)

    assert_receive {:second_batch, [second_blob]}
    assert second_blob == Kernel.encode(second)
    assert sent["peer"] == MapSet.new([first_id, second_id])
  end

  test "wire input is incremental and bounded output drains partially without prefix copies" do
    wire = Runtime.wire_message(0, "hello") <> Runtime.wire_message(1, "world")
    decoder = %WireDecoder{}

    {messages, decoder} = WireDecoder.feed(decoder, binary_part(wire, 0, 3))
    assert messages == []

    {messages, decoder} = WireDecoder.feed(decoder, binary_part(wire, 3, 5))
    assert messages == []

    {messages, decoder} =
      WireDecoder.feed(decoder, binary_part(wire, 8, byte_size(wire) - 8))

    assert messages == [{0, "hello"}, {1, "world"}]
    assert decoder.buffer == <<>>

    link = OutLink.new(byte_size(wire))
    {:ok, link} = OutLink.enqueue(link, 0, "hello")
    {:ok, link} = OutLink.enqueue(link, 1, "world")
    assert {:error, :full, ^link} = OutLink.enqueue(link, 1, "overflow")

    {prefix, link} = OutLink.take(link, 3)
    assert link.pending == byte_size(wire) - 3
    {tail, link} = OutLink.take(link, 10_000)

    assert prefix <> tail == wire
    assert link.pending == 0
    assert :queue.is_empty(link.chunks)
  end

  defp offer(role, scope, target, value \\ nil) do
    %Atom{kind: :offer, role: role, scope: scope, target: target, value: value}
  end

  defp need(role, scope, target, effect) do
    %Atom{kind: :need, role: role, scope: scope, target: target, effect: effect}
  end
end
